### PR TITLE
mds/MDSDaemon: fix asok 'exit' and 'respawn' commands

### PR DESCRIPTION
**Pull request by liewegas**
Signed-off-by: Sage Weil <sage@redhat.com>

### CODE REVIEW

File: src/mds/MDSDaemon.cc [hunk 1]
```diff
@@ -144,21 +144,25 @@ void MDSDaemon::asok_command(
...
     r = 0;
-    std::thread t([this](){
+    auto t = new std::thread([this](){
 		    // Wait a little to improve chances of caller getting
 		    // our response before seeing us disappear from mdsmap
 		    sleep(1);
+		    std::lock_guard l(mds_lock);
 		    suicide();
 		  });
+    (void)t;
```
Comment (batrick) [inline src/mds/MDSDaemon.cc]: Would it not be better to do:

`t.detach();` instead?

File: src/mds/MDSDaemon.cc [hunk 2]
```diff
@@ -144,21 +144,25 @@ void MDSDaemon::asok_command(
...
     r = 0;
-    std::thread t([this](){
+    auto t = new std::thread([this](){
 		    // Wait a little to improve chances of caller getting
 		    // our response before seeing us disappear from mdsmap
 		    sleep(1);
+		    std::lock_guard l(mds_lock);
 		    respawn();
 		  });
+    (void)t;
```
Comment (batrick) [inline src/mds/MDSDaemon.cc]: ditto

File: src/mds/MDSDaemon.cc (hunk 1 above)
Comment (liewegas) [reply src/mds/MDSDaemon.cc]: I think ``auto t`` should make ``t`` a ``std::thread*``, not a smart pointer?

Comment (liewegas) [reply src/mds/MDSDaemon.cc]: oh, std::thread::detach(), gotcha

**Comment by liewegas** [comment]
retest this please
